File: DataStructutreAndAlgorithm/Algorithm/02_Sorting.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/* Quick Sort */
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;
    return i + 1;
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

/* Merge Sort */
void merge(int arr[], int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;
    
    int *L = (int *)malloc(n1 * sizeof(int));
    int *R = (int *)malloc(n2 * sizeof(int));
    
    memcpy(L, &arr[left], n1 * sizeof(int));
    memcpy(R, &arr[mid + 1], n2 * sizeof(int));
    
    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) arr[k++] = L[i++];
        else arr[k++] = R[j++];
    }
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
    
    free(L);
    free(R);
}

void mergeSort(int arr[], int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

void printArray(int arr[], int n) {
    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d", arr[i]);
        if (i < n - 1) printf(", ");
    }
    printf("]\n");
}

int isSorted(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        if (arr[i] < arr[i - 1]) return 0;
    }
    return 1;
}

int main() {
    printf("===== SORTING ALGORITHMS =====\n\n");
    
    int arr1[] = {64, 34, 25, 12, 22, 11, 90, 88, 45, 50};
    int arr2[] = {64, 34, 25, 12, 22, 11, 90, 88, 45, 50};
    int n = sizeof(arr1) / sizeof(arr1[0]);
    
    printf("Original: ");
    printArray(arr1, n);
    printf("\n");
    
    /* Quick Sort */
    printf("--- QUICK SORT ---\n");
    clock_t start = clock();
    quickSort(arr1, 0, n - 1);
    clock_t end = clock();
    printf("Sorted: ");
    printArray(arr1, n);
    printf("Valid: %s\n", isSorted(arr1, n) ? "YES" : "NO");
    printf("Time: %.3f ms\n\n", ((double)(end - start)) / CLOCKS_PER_SEC * 1000);
    
    /* Merge Sort */
    printf("--- MERGE SORT ---\n");
    start = clock();
    mergeSort(arr2, 0, n - 1);
    end = clock();
    printf("Sorted: ");
    printArray(arr2, n);
    printf("Valid: %s\n", isSorted(arr2, n) ? "YES" : "NO");
    printf("Time: %.3f ms\n\n", ((double)(end - start)) / CLOCKS_PER_SEC * 1000);
    
    /* Edge cases */
    printf("--- EDGE CASES ---\n");
    
    int sorted[] = {1, 2, 3, 4, 5};
    int n2 = 5;
    printf("Already sorted: ");
    printArray(sorted, n2);
    quickSort(sorted, 0, n2 - 1);
    printf("After sort: ");
    printArray(sorted, n2);
    printf("\n");
    
    int reverse[] = {5, 4, 3, 2, 1};
    printf("Reverse sorted: ");
    printArray(reverse, n2);
    mergeSort(reverse, 0, n2 - 1);
    printf("After sort: ");
    printArray(reverse, n2);
    printf("\n");
    
    /* Performance test */
    printf("--- PERFORMANCE TEST ---\n");
    int size = 50000;
    int *perfArr1 = (int *)malloc(size * sizeof(int));
    int *perfArr2 = (int *)malloc(size * sizeof(int));
    
    srand(42);
    for (int i = 0; i < size; i++) {
        perfArr1[i] = rand() % 10000;
        perfArr2[i] = perfArr1[i];
    }
    
    start = clock();
    quickSort(perfArr1, 0, size - 1);
    end = clock();
    double quickTime = ((double)(end - start)) / CLOCKS_PER_SEC * 1000;
    
    start = clock();
    mergeSort(perfArr2, 0, size - 1);
    end = clock();
    double mergeTime = ((double)(end - start)) / CLOCKS_PER_SEC * 1000;
    
    printf("Sorting %d elements:\n", size);
    printf("  Quick Sort: %.3f ms\n", quickTime);
    printf("  Merge Sort: %.3f ms\n\n", mergeTime);
    
    free(perfArr1);
    free(perfArr2);
    
    printf("=== SUMMARY ===\n");
    printf("Quick Sort:  O(n log n) avg, O(n²) worst, O(log n) space\n");
    printf("Merge Sort:  O(n log n) always, O(n) space\n");
    printf("Network App: QoS priority, routing metrics\n");
    
    return 0;
}

File: DataStructutreAndAlgorithm/Algorithm/02_Sorting.java
import java.util.Arrays;
import java.util.Random;

public class SortingDemo {
    
    /* Quick Sort */
    static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        return i + 1;
    }
    
    static void quickSort(int[] arr) {
        if (arr != null && arr.length > 0) {
            quickSort(arr, 0, arr.length - 1);
        }
    }
    
    static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }
    
    /* Merge Sort */
    static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;
        
        int[] L = new int[n1];
        int[] R = new int[n2];
        
        System.arraycopy(arr, left, L, 0, n1);
        System.arraycopy(arr, mid + 1, R, 0, n2);
        
        int i = 0, j = 0, k = left;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) arr[k++] = L[i++];
            else arr[k++] = R[j++];
        }
        while (i < n1) arr[k++] = L[i++];
        while (j < n2) arr[k++] = R[j++];
    }
    
    static void mergeSort(int[] arr) {
        if (arr != null && arr.length > 0) {
            mergeSort(arr, 0, arr.length - 1);
        }
    }
    
    static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = left + (right - left) / 2;
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
            merge(arr, left, mid, right);
        }
    }
    
    static boolean isSorted(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] < arr[i - 1]) return false;
        }
        return true;
    }
    
    public static void main(String[] args) {
        System.out.println("===== SORTING ALGORITHMS =====\n");
        
        int[] arr1 = {64, 34, 25, 12, 22, 11, 90, 88, 45, 50};
        int[] arr2 = {64, 34, 25, 12, 22, 11, 90, 88, 45, 50};
        
        System.out.println("Original: " + Arrays.toString(arr1) + "\n");
        
        /* Quick Sort */
        System.out.println("--- QUICK SORT ---");
        long start = System.nanoTime();
        quickSort(arr1);
        long quickTime = System.nanoTime() - start;
        System.out.println("Sorted: " + Arrays.toString(arr1));
        System.out.println("Valid: " + (isSorted(arr1) ? "YES" : "NO"));
        System.out.println("Time: " + String.format("%.3f ms\n", quickTime / 1_000_000.0));
        
        /* Merge Sort */
        System.out.println("--- MERGE SORT ---");
        start = System.nanoTime();
        mergeSort(arr2);
        long mergeTime = System.nanoTime() - start;
        System.out.println("Sorted: " + Arrays.toString(arr2));
        System.out.println("Valid: " + (isSorted(arr2) ? "YES" : "NO"));
        System.out.println("Time: " + String.format("%.3f ms\n", mergeTime / 1_000_000.0));
        
        /* Edge cases */
        System.out.println("--- EDGE CASES ---");
        
        int[] sorted = {1, 2, 3, 4, 5};
        System.out.println("Already sorted: " + Arrays.toString(sorted));
        quickSort(sorted);
        System.out.println("After sort: " + Arrays.toString(sorted) + "\n");
        
        int[] reverse = {5, 4, 3, 2, 1};
        System.out.println("Reverse sorted: " + Arrays.toString(reverse));
        mergeSort(reverse);
        System.out.println("After sort: " + Arrays.toString(reverse) + "\n");
        
        int[] dups = {3, 1, 4, 1, 5, 9, 2, 6, 5};
        System.out.println("With duplicates: " + Arrays.toString(dups));
        quickSort(dups);
        System.out.println("After sort: " + Arrays.toString(dups) + "\n");
        
        /* Performance test */
        System.out.println("--- PERFORMANCE TEST ---");
        int[] sizes = {10000, 100000, 1000000};
        
        for (int size : sizes) {
            Random rand = new Random(42);
            int[] quickArr = new int[size];
            int[] mergeArr = new int[size];
            
            for (int i = 0; i < size; i++) {
                quickArr[i] = rand.nextInt(10000);
                mergeArr[i] = quickArr[i];
            }
            
            start = System.nanoTime();
            quickSort(quickArr);
            quickTime = System.nanoTime() - start;
            
            start = System.nanoTime();
            mergeSort(mergeArr);
            mergeTime = System.nanoTime() - start;
            
            System.out.println("Size: " + size);
            System.out.println("  Quick Sort: " + String.format("%.3f ms", quickTime / 1_000_000.0));
            System.out.println("  Merge Sort: " + String.format("%.3f ms\n", mergeTime / 1_000_000.0));
        }
        
        /* Network Application: QoS Scheduling */
        System.out.println("--- NETWORK APPLICATION: QoS Scheduling ---");
        String[] packets = {"VoIP(priority=5)", "BulkData(priority=1)", 
                           "Video(priority=4)", "Email(priority=2)"};
        int[] priorities = {5, 1, 4, 2};
        
        System.out.println("Before sorting:");
        for (int i = 0; i < packets.length; i++) {
            System.out.println("  " + packets[i]);
        }
        
        /* Simple sort by priority */
        for (int i = 0; i < priorities.length; i++) {
            for (int j = i + 1; j < priorities.length; j++) {
                if (priorities[j] > priorities[i]) {
                    int tempP = priorities[i];
                    priorities[i] = priorities[j];
                    priorities[j] = tempP;
                    
                    String tempPk = packets[i];
                    packets[i] = packets[j];
                    packets[j] = tempPk;
                }
            }
        }
        
        System.out.println("\nAfter sorting by priority (high to low):");
        for (int i = 0; i < packets.length; i++) {
            System.out.println("  " + packets[i]);
        }
        System.out.println("\nHigh priority packets processed first!\n");
        
        System.out.println("=== SUMMARY ===");
        System.out.println("Quick Sort:  O(n log n) avg, O(n²) worst, O(log n) space");
        System.out.println("Merge Sort:  O(n log n) always, O(n) space");
        System.out.println("Network App: QoS priority scheduling, routing metrics");
    }
}
